### DISCUSSION TITLE
Adding New Control Interfaces

### DISCUSSION THREAD

**Discussion by wildseyed** [probonopd/MiniDexed#117] [Ideas]
I went looking for an LCD and rotary encoder this morning at Microcenter, but no luck on the encoder. I found an Adafruit Inverse RGB 16x2 LCD Hat with a button interface, and this has me thinking about what would be needed to facilitate the use of alternate hardware as an interface, but also things like OSC and the need for a HAL or maybe just a MIDI bridge to OSC.

**Comment by probonopd** [comment]
Hello @wildseyed.

KY-040 rotary encoders should be easy to order online. Should be around 2 USD. The Adafruit Inverse RGB 16x2 LCD Hat is not usable (at least not without taking it apart). What you need is a much simpler "bare" 16x2 LCD (not i2c).  Should be around 3 USD.

Note that you can get started without the display and rotary encoder, but they are indeed very useful if you don't want to work by editing files on the SD card.

OSC is not planned. The DX7 and friends are MIDI instruments.

**Reply by wildseyed** [reply]
Hi @probonopd 

Yes, I did not mean to imply that I could not source the supported components. Only that the LCD Hat (and OSC) would provide more streamlined interfaces for working with the patch parameters, not needing breadboards and such. With a hardware abstraction layer, adding support for any type of control input would be simpler. As for DX7 being a MIDI instrument, this would not be a reason for excluding OSC from the roadmap, although I certainly understand that someone would have to take ownership of the effort. OSC works very well with MIDI instruments. I think I read that SYSEX command support is planned. If this is the case, bridging OSC to SYSEX should be possible. It is something I am willing to explore further. I first need to get familiar with Circle. Did not know this was a thing. For now, I am having a blast playing with this. Thanks!

**Reply by probonopd** [reply]
Yes, once https://github.com/probonopd/MiniDexed/issues/82 is implemented external devices (e.g., an app running on a tablet, or Dexed running on another computer) could be used to control MiniDexed. What would be the advantage of using OSC over MIDI?

https://github.com/probonopd/MiniDexed/issues/63 would even allow us to use knobs, sliders, and buttons on existing MIDI hardware that can't even send the DX7 sysex.

By the way, I am not using breadboards. Just the bare components with some DuPont wires and a liberal amount of hot glue. Maybe we will design a simple PCB for home soldering at some point.

**Reply by wildseyed** [reply]
I don't think it's a question of OSC versus MIDI. The two protocols work in tandem to facilitate device configuration and use. For OSC, what I have in mind is using TouchOSC in a manner similar to what [Matrix](https://www.matrixsynth.com/2011/02/touchosc-dx7-editor-for-missing-link.html?m=1) posted for use with The Missing Link, which I think is a hardware MIDI bridge, no longer in production. There are plenty of OSC to MIDI Bridge software examples. That template is no longer available, but implementing something like it in TouchOSC is straightforward.

**Reply by probonopd** [reply]
I am not super familiar with TouchOSC but I could imagine that once https://github.com/probonopd/MiniDexed/issues/82 is implemented TouchOSC should be able to control MiniDexed?

**Comment by mazoqui** [comment]
I'm new to the technologies you used to create MiniDexed and I confess that I fell in love with your project. Despite my little knowledge, I managed to run it on an RP v1 and was impressed with thes surprisely performance. 
I don't have any UI connected to my board (no rotate encoder/lcd display), but I have connected an HDMI monitor to my board. It's a bit confusing for me to fully enjoy it, without interacting with via UI. Anyway, I took a look at the code and noticed that there is a class that implements a PC keyboard, inheriting from a MIDI device. Would It be a standard breaking, whether the same instance (pckeyboard)  generates events similar to those handled by the method CUserInterface::EncoderEventHandler (eg: ">" would generates the  CUIMenu::MenuEventStepUp event, etc...)?
As I mentioned I'm new to all this and I wouldn't even dare try to add some functionality without listening to more experienced people.
Thank you for such wonderful job. Congrats!

**Reply by probonopd** [reply]
So you would rather like to use a PC keyboard than a Rotary Encoder? You would still need the LCD. The text on the HDMI screen is really just debug output for developers at this point. (I think who wants to actually use a HDMI screen might be better of which a DAW like Reaper running Dexed as a plugin anyway...)

**Reply by mazoqui** [reply]
Hi @probonopd,
thank you. 
My currently setup is just an ordinary keyboard connected through midi usb to RPpi board, with no computer at all. Anyway, thank you for the advice. I will take a look a look on Dexed too

**Reply by probonopd** [reply]
If your keyboard can send MIDI control change messages, then you can control MiniDexed this way, e.g., to change voices and banks.
